File: python/raytracer/objects/Disc.py
import numpy as np
from ..Material import Material
from ..Ray import Ray
from .Plane import Plane


class Disc(Plane):
    def __init__(self, position: [float], normal: [float], radius: float, material: Material):
        super().__init__(position, normal, material)
        self.radius = radius

    def get_intersection(self, ray: Ray):
        intersection = super().get_intersection(ray)

        if intersection is not None:
            plane_vector = intersection.hit_point - self.position

            if np.linalg.norm(plane_vector) <= self.radius:
                return intersection

        return None

File: src/objects/Disc.cpp
#include "raytracer/objects/Disc.h"


Disc::Disc(const Eigen::Vector3d& position, const Eigen::Vector3d& normal, const double& radius, const Material& material)
: Plane(position, normal, material)
, radius(radius)
{ }


bool Disc::get_intersection(Intersection &intersection, const Ray &ray) {
    if (Plane::get_intersection(intersection, ray)) {
        Eigen::Vector3d plane_vector = intersection.get_hit_point() - this->position;

        return plane_vector.norm() <= this->get_radius();
    }
    return false;
}


double Disc::get_radius() const {
    return this->radius;
}
